google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java

```diff
@@ -349,33 +388,47 @@ public StreamingCall executeQuery(
   @Override
   public Transaction beginTransaction(
       BeginTransactionRequest request, @Nullable Map<Option, ?> options) throws SpannerException {
-    throw new UnsupportedOperationException("Not implemented yet.");
+    GrpcCallContext context = GrpcCallContext.createDefault()
+        .withChannelAffinity(Option.CHANNEL_HINT.getLong(options).intValue())
+        .withExtraHeaders(metadataProvider.newExtraHeaders(request.getSession(), projectName));
```

Comment:
Can you extract out this piece of code into a helper method. It is repeated in multiple places.